dapp_runner/_util.py

```diff
@@ -11,23 +12,37 @@
 from yapapi import __version__ as yapapi_version
 
 
-def get_free_port(range_start: int = 8080, range_end: int = 9090) -> int:
-    """Get the first available port on localhost within the specified range.
+@lru_cache(maxsize=1)
+def _free_port_generator(
+    range_start: int = 8080, range_end: int = 9090
+) -> Generator[int, None, None]:
+    """Yield the first available port on localhost within the specified range.
 
     The range is inclusive on both sides (i.e. `range_end` will be included).
     Raises `RuntimeError` when no free port could be found.
+
+    `lru_cache` prevents race condition on single dapp-runner instance when reserving free ports.
     """
+    return_port = None
     for port in range(range_start, range_end + 1):
         with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
             try:
                 s.bind(("", port))
-                return port
+                return_port = port
             except OSError:
                 pass
-
+        if return_port is not None:
+            yield return_port
+            return_port = None
+            
     raise RuntimeError(f"No free ports found. range_start={range_start}, range_end={range_end}")
 
 
+def get_free_port() -> int:
```

Comment:
it would still be good if `get_free_port` had a start + end range, since we'd like to keep the range configurable